### ISSUE TITLE
[Roadmap] Discussion of future features and enhancements

### ISSUE THREAD

**Issue by jfjallid** [jfjallid/go-smb#19]
The development of this library has primarily been driven by the needs of the various tools I've created such as [go-secdump](https://github.com/jfjallid/go-secdump), [go-shareenum](https://github.com/jfjallid/go-shareenum), and [go-cmloot](https://github.com/jfjallid/go-cmloot). It started as a pure SMB library but quickly included some DCERPC and more recently Kerberos support. A long term goal is to add support for LDAP and extend the Kerberos integration to be able to support more of the actions that can be performed using Impacket.

I do not currently have any detailed roadmap in mind but if anyone have ideas I'm open to suggestions.
Feel free to add comments here to suggest new features/enhancements or open new issues directly with a proposal for changes.

**Comment by iamalsaher** [comment]
Yup, I want to do the same. Have a golang library which behaves like impacket, broken down into different core modules, which can be used with each other. 

I think the primary goals are to use
1. SMB based features. Simple API to perform SMB tasks while providing options for configurability.
2. Credentials and Authentication. A simple API for authentication which can be used across multiple modules, supporting auth via ntlm, kerberos, ekeys, certs, logged in user(sspi bind) etc. Availibility to parse each message, ticket, convert to multiple trackable formats.
3. Execution options coupled on top of these wmiexec, psexec, dcom, scheduler etc
4. Relaying, NTLM and LLMNR

I think creating a right architecture for this is crucial, and will allow people to utilise the library in offensive, as well as general form.

**Comment by jfjallid** [comment]
> Recently, I have been exploring [Alex's SSPI Library](https://github.com/alexbrainman/sspi/) which allows me to bind sessions to the currently logged in user and perform activities in their security context. I think having a support for this in go-smb would be great. Just in case it is there, I have really not been able to find it

Currently no support for that but it would be a welcome addition. The tools using this library have been designed to run remotely, typically on a Linux machine, so no attempts at evading EDRs locally or reusing local creds.

**Comment by iamalsaher** [comment]
> There are some major changes that would be required to be more versatile like Impacket. For instance, the DCERPC support is currently implemented to run over SMB only but for some things it would be more useful to be able to select transport between SMB and running directly on top of TCP. This library also lacks some of the flexibility available in Impacket to modify and hook requests to support custom use cases.
> 
> There is currently some limited support available for NTLM relaying but nothing for LLMNR poisoning.
> 
> Additional methods of authentication would be nice to have. Currently NTLM and limited Kerberos auth is supported.
> 
> Regarding execution tools like wmiexec, psexec, etc. I think it might be better to keep those tools as separate repos which rely on go-smb for the transport and authentication and keep the tool specific logic local to those repos, such as the upload of a service binary, the CLI interface to interact with the service etc.

Agree on your points completely, I have also been on the lookout for other libraries, which can help support our usecases 
1. https://github.com/oiweiwei/go-msrpc
2. https://github.com/Amzza0x00/go-impacket
3. https://github.com/RedTeamPentesting/pretender

Can you suggest on how we should move forward.

**Comment by iamalsaher** [comment]
> > Recently, I have been exploring [Alex's SSPI Library](https://github.com/alexbrainman/sspi/) which allows me to bind sessions to the currently logged in user and perform activities in their security context. I think having a support for this in go-smb would be great. Just in case it is there, I have really not been able to find it
> 
> Currently no support for that but it would be a welcome addition. The tools using this library have been designed to run remotely, typically on a Linux machine, so no attempts at evading EDRs locally or reusing local creds.

I will try to see if I can natively integrate the sspi client in the library

**Comment by jfjallid** [comment]
In the latest release v0.7.0 I've created the transport interface and implemented support for DCERPC over TCP in addition to SMB so it is now possible to choose which transport to use as well as to lookup which dynamic port number a given service is bound to using the Endpoint Mapper. So it should now be possible to add support for services that require TCP transport.
